EditingToolkitSidebar: add tracking events

autumnfjeld opened the pull request:
### Changes proposed in this Pull Request

Adds tracking events for
* When the sidebar is opened
* When the user goes back to the dashboard, closing the editor
* When the user navigates to an existing post/page
* When the user clicks on "add a new post/page"
* When sidebar is dismissed

### Testing instructions

* Add diff D49173-code to your sandbox for testing
* open dev tools  and paste this in console: localStorage.setItem('debug', 'calypso:analytics');   this will allow verification of analytics in the dev tools console
* Some events can't be viewed in the console due to a page load, in this case verify the even in Tracks: 

**Open sidebar**
    * click the (W)
    * Verify `calypso:analytics Recording event "calypso_editor_sidebar_open" ` in dev tools console

**Close editor / back to dashboard** 
    * When sidebar is open
    * click All Posts (or All Pages)
    *Verify `calypso:analytics Record event "calypso_editor_sidebar_editor_close` in dev tools console

**Navigate to an existing post/page**
    * When sidebar is open
    * click the Post or Page name in the sidebar list
    * Verify event in tracks: calypso_editor_sidebar_item_edit

**Click on add new post/page button**
    * When sidebar is open
    * click Add new page/post
    * Verify in tracks calypso_editor_sidebar_item_add

**Sidebar is dismissed/closed**
    * When sidebar is open
    * press esc or click (W) or click anywhere outside sidebar
    * Verify `calypso:analytics Record event "calypso_editor_sidebar_dismiss"` in dev tools console

### Issue
Fixes # https://github.com/Automattic/wp-calypso/issues/44907

autumnfjeld commented:
@p-jackson or @ramonjd  (I thought there was a @ganon type of tag, but it is not showing up. 🧐  So just tagging directly.)  I need to add tracking to the (W) in the header, to track 'sidebar open'. 
 This button is called `ToggleSidebarButton`, but as far as I can tell, it is not a toggle. It only opens the sidebar. It does not close the sidebar.  I just wanted to verify that I am correct about that. (This is kind of a confirmation of the chat we had about this last week, the close sidebar (W) button is a different button that lives in editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx)

https://github.com/Automattic/wp-calypso/blob/94fb6b47a55791f413f64b61babf607581954ee7/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/toggle-sidebar-button/index.tsx

The only tracking that would go on the  `ToggleSidebarButton` is an 'sidebar open' tracking event.  Ya?

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/create-page/index.tsx:
```diff
@@ -43,13 +44,18 @@ export default function CreatePage( { postType }: Props ) {
 	);
 
+	const trackEvent = () => {
+		recordTracksEvent( `calypso_editor_sidebar_${ postType.slug }_add` );
```
I'm not sure whether it's best to to track the `_add` event as two seperate events (one for posts and one for pages) or as one event. The advantage of a single event is it makes it easier to determine how often users are creating _anything_ using the sidebar, and I _think_ that's what we're more interested in.

We can still track the post type just in case with an event param:
```suggestion
		recordTracksEvent( 'calypso_editor_sidebar_add', { post_type: postType.slug } );
```

p-jackson replied:
What about `calypso_editor_sidebar_item_add`?

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-item/index.tsx:
```diff
@@ -38,12 +39,16 @@ const NavItem = forwardRef< HTMLLIElement, NavItemProps >(
 			postType.slug
 		);
+		const trackEvent = () => {
+			recordTracksEvent( `calypso_editor_sidebar_${ postType.slug }_open` );
```
Same comment here as in the `_add` event, maybe post type should be an event param.

p-jackson replied:
What about `calypso_editor_sidebar_item_edit`. Since "open" might mean to open the sidebar.

autumnfjeld replied:
Ah yes, edit much more specific!

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx:
```diff
@@ -113,6 +114,7 @@ function WpcomBlockEditorNavSidebar() {
 	const handleClose = ( e: React.MouseEvent ) => {
 		if ( hasAction( 'a8c.wpcom-block-editor.closeEditor' ) ) {
 			e.preventDefault();
+			recordTracksEvent( 'calypso_editor_sidebar_close' );
```
```suggestion
			recordTracksEvent( 'calypso_editor_sidebar_close_editor' );
```

"close editor" might help distinguish it from closing the sidebar. Or maybe it should be `_editor_close` since we often put the verb at the end in track event names.

autumnfjeld replied:
Yes, I see what you mean.  'calypso_editor_sidebar_editor_close' follows the pattern, and then intention.  Probably confusing regardless. :). But I suppose that is where https://mc.a8c.com/tracks/events/register/ comes in, to add descriptions & clairty.

p-jackson reviewed (commented):
I haven't run it yet but the code is looking good 👍 

It's missing an event when the sidebar is dismissed, but I'm sure that's on it's way.

The other way to send events when the sidebar toggles would be to do it in the Redux layer. That way even a hypothetical feature that toggles the sidebar (even if it's something outside the sidebar UI) wouldl still sent an event sent. On the other hand maybe we don't want that: tracks events are supposed to be user action focused, so we should be able to map them 1-to-1 with a user action (like clicking a button).
Just something to consider, not something that we should necessarily do.

autumnfjeld commented:
@p-jackson 
> It's missing an event when the sidebar is dismissed, but I'm sure that's on it's way.

I don't think tracking for "sidebar is dismissed"  is specified here: https://github.com/Automattic/wp-calypso/issues/44907. but I went ahead and added it.  I assume it would be useful in the case that user's are opening/closing it without using any functionality inside?  As in that may indicate confusion?

I used ` 'calypso_editor_sidebar_dismiss'`.  Is that the verb we want? Or do we want ` 'calypso_editor_sidebar_close'` ?
https://github.com/Automattic/wp-calypso/pull/45429/commits/d547d12e947501170c1ba1b1c05697df2ba0eb5b

p-jackson commented:
Now that you mention it, I don't know why I assumed there should be a dismiss event 😅. Maybe it's just because when doing other tracking (like in gutenboarding) we added open/dismiss events in pairs.

I can't think of a super compelling reason to have a dismiss event now. Feel free to leave it or take it out if you want. Leaving it in to track whether users are opening the sidebar without clicking anything, or even just for symmetry, is reasonable.

> I used `'calypso_editor_sidebar_dismiss'`. Is that the verb we want?

If you keep it then personally I prefer dismiss.

autumnfjeld commented:
@p-jackson I'm ready for a final ✅ .  I've addressed all your comments.

One test is failing, but I don't think that is from my work.  Seems likely that it is an e2e test issue (with a css selector).

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx:
```diff
@@ -136,6 +138,7 @@ function WpcomBlockEditorNavSidebar() {
 			  );
 
 	const dismissSidebar = () => {
+		recordTracksEvent( 'calypso_editor_sidebar_dismiss' );
```
The event should probably inside the if statement (it's only a true dismissal if `isOpen && !isClosing`), but I can't reproduce this in practice.

p-jackson reviewed (approved):
Great! Events fire as expected, and I see them appear in the backend 👍 

:shipit: 

> Some events can't be viewed in the console due to a page load, in this case verify the even in Tracks

The other way around this is to enable the "preserve log" setting so the console isn't cleared across page loads.

![Screen Shot on 2020-09-09 at 10:55:48](https://user-images.githubusercontent.com/1500769/92535337-17174b80-f28b-11ea-899d-b208267a6ef8.png)

I usually only toggle it on when I want it, because the console gets full pretty fast!

autumnfjeld commented:
@Automattic/ganon  Looking for a quick review on each of the registration PRs below.  😃 
Event Registration:
389-gh-Automattic/tracks-events-registration
390-gh-Automattic/tracks-events-registration
391-gh-Automattic/tracks-events-registration
